File: add_product_tags.py
#!/usr/bin/env python3
"""Add product tags to recipe markdown files based on Chang's product links in ingredients."""

import os
import re
import glob

RECIPES_DIR = os.path.join(os.path.dirname(__file__), "recipes")

def extract_product_tags_from_content(content):
    """Extract Chang's product tag names from product links in the recipe content."""
    # Find all title attributes from Chang's product links
    titles = re.findall(r'href="/products/[^"]+"\s*(?:target="[^"]*"\s*)?(?:rel="[^"]*"\s*)?>([^<]+)</a>', content)
    # Also try the other order: title first
    titles += re.findall(r'title="(Chang\'s [^"]+)"', content)

    tags = set()
    for title in titles:
        title = title.strip()
        if not title.startswith("Chang's "):
            continue
        # Strip "Chang's " prefix
        tag = title[len("Chang's "):]
        # Normalize whitespace
        tag = " ".join(tag.split())
        # Remove size suffixes like "(280ml)", "(150ml)" etc.
        tag = re.sub(r'\s*\(\d+ml\)\s*$', '', tag).strip()
        # Fix known typo
        tag = tag.replace("Soy Sauce Lighte", "Soy Sauce Light")
        if tag:
            tags.add(tag)
    return tags


def get_existing_tags(content):
    """Extract existing tags from the YAML frontmatter."""
    # Match the tags section
    tags_match = re.search(r'^tags:\s*\n((?:\s+-\s+"[^"]*"\s*\n)*)', content, re.MULTILINE)
    if not tags_match:
        return set()
    tags_block = tags_match.group(1)
    tags = re.findall(r'^\s+-\s+"([^"]+)"', tags_block, re.MULTILINE)
    return set(tags)


def add_tags_to_file(filepath, new_tags):
    """Add missing product tags to the end of the tags section."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    existing = get_existing_tags(content)
    missing = sorted(new_tags - existing)

    if not missing:
        return False

    # Find the end of the tags section (just before closing ---)
    # The tags section ends either at a blank line followed by --- or at ---
    # We need to insert after the last tag entry

    # Find the tags block and append to it
    # Pattern: find last tag line and insert after it
    tags_section = re.search(r'(^tags:\s*\n(?:\s+-\s+"[^"]*"\s*\n)*)', content, re.MULTILINE)
    if not tags_section:
        print(f"  WARNING: no tags section found in {os.path.basename(filepath)}")
        return False

    tags_block = tags_section.group(1)
    new_lines = "".join(f'  - "{t}"\n' for t in missing)
    new_tags_block = tags_block + new_lines

    new_content = content[:tags_section.start()] + new_tags_block + content[tags_section.end():]

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(new_content)

    return True


def process_recipes():
    files = sorted(glob.glob(os.path.join(RECIPES_DIR, "*.md")))
    updated = 0
    skipped = 0

    for filepath in files:
        name = os.path.basename(filepath)
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        product_tags = extract_product_tags_from_content(content)
        if not product_tags:
            skipped += 1
            continue

        existing = get_existing_tags(content)
        missing = product_tags - existing
        if not missing:
            skipped += 1
            continue

        print(f"{name}: adding {sorted(missing)}")
        if add_tags_to_file(filepath, product_tags):
            updated += 1

    print(f"\nDone: {updated} files updated, {skipped} files unchanged.")


if __name__ == "__main__":
    process_recipes()

File: add_product_tags.js
#!/usr/bin/env node
// Add product tags to recipe markdown files based on Chang's product links in ingredients.

const fs = require('fs');
const path = require('path');

const RECIPES_DIR = path.join(__dirname, 'recipes');

function extractProductTags(content) {
  const tags = new Set();

  // Match <a ... href="/products/..." ... title="Chang's ..."> in any attribute order.
  // We need to find anchors whose href is a /products/ path AND has a Chang's title.
  const anchorRe = /<a\s+[^>]*href="\/products\/[^"]*"[^>]*>/gi;
  let anchor;
  while ((anchor = anchorRe.exec(content)) !== null) {
    const titleMatch = /title="(Chang's [^"]+)"/.exec(anchor[0]);
    if (!titleMatch) continue;

    let tag = titleMatch[1].slice("Chang's ".length).trim();
    // Normalize whitespace
    tag = tag.replace(/\s+/g, ' ').trim();
    // Remove size suffixes like "(280ml)"
    tag = tag.replace(/\s*\(\d+ml\)\s*$/, '').trim();
    // Fix known typos
    tag = tag.replace('Soy Sauce Lighte', 'Soy Sauce Light');
    if (tag) tags.add(tag);
  }

  // Also handle reversed attribute order: title first, then href
  const anchorRe2 = /<a\s+[^>]*title="(Chang's [^"]+)"[^>]*href="\/products\/[^"]*"[^>]*>/gi;
  while ((anchor = anchorRe2.exec(content)) !== null) {
    let tag = anchor[1].slice("Chang's ".length).trim();
    tag = tag.replace(/\s+/g, ' ').trim();
    tag = tag.replace(/\s*\(\d+ml\)\s*$/, '').trim();
    tag = tag.replace('Soy Sauce Lighte', 'Soy Sauce Light');
    if (tag) tags.add(tag);
  }

  return tags;
}

function getExistingTags(content) {
  const existing = new Set();
  // Match the full tags section, allowing for blank lines within it
  const tagsRe = /^tags:\s*\n([\s\S]*?)^---/m;
  const match = tagsRe.exec(content);
  if (!match) return existing;

  const lineRe = /^\s+-\s+"([^"]+)"/gm;
  let m;
  while ((m = lineRe.exec(match[1])) !== null) {
    existing.add(m[1]);
  }
  return existing;
}

function addTagsToFile(filepath, productTags) {
  const content = fs.readFileSync(filepath, 'utf8');
  const existing = getExistingTags(content);
  const missing = [...productTags].filter(t => !existing.has(t)).sort();

  if (missing.length === 0) return false;

  // Find the last tag line in the tags section and insert after it.
  // Use a regex that stops at the closing --- to find the last tag line.
  const tagsBlockRe = /^tags:\s*\n([\s\S]*?)^---/m;
  const blockMatch = tagsBlockRe.exec(content);
  if (!blockMatch) {
    console.log(`  WARNING: no tags section in ${path.basename(filepath)}`);
    return false;
  }

  // Find the position of the last tag line within the block
  const blockStart = blockMatch.index + 'tags:\n'.length;
  const blockContent = blockMatch[1];

  // Find the last tag entry in the block
  const lastTagRe = /^\s+-\s+"[^"]*"\s*\n/gm;
  let lastTagMatch = null;
  let m;
  while ((m = lastTagRe.exec(blockContent)) !== null) {
    lastTagMatch = m;
  }

  if (!lastTagMatch) {
    console.log(`  WARNING: no tag entries found in ${path.basename(filepath)}`);
    return false;
  }

  const insertPos = blockStart + lastTagMatch.index + lastTagMatch[0].length;
  const newLines = missing.map(t => `  - "${t}"\n`).join('');
  const newContent = content.slice(0, insertPos) + newLines + content.slice(insertPos);

  fs.writeFileSync(filepath, newContent, 'utf8');
  return true;
}

const files = fs.readdirSync(RECIPES_DIR)
  .filter(f => f.endsWith('.md'))
  .sort()
  .map(f => path.join(RECIPES_DIR, f));

let updated = 0, skipped = 0;

for (const filepath of files) {
  const name = path.basename(filepath);
  const content = fs.readFileSync(filepath, 'utf8');
  const productTags = extractProductTags(content);

  if (productTags.size === 0) {
    skipped++;
    continue;
  }

  const existing = getExistingTags(content);
  const missing = [...productTags].filter(t => !existing.has(t));

  if (missing.length === 0) {
    skipped++;
    continue;
  }

  console.log(`${name}: adding [${missing.sort().join(', ')}]`);
  if (addTagsToFile(filepath, productTags)) updated++;
}

console.log(`\nDone: ${updated} files updated, ${skipped} files unchanged.`);
